### DISCUSSION TITLE
Want to contribute to the code, but where do I start?

### DISCUSSION THREAD

**Discussion by HuangStomach** [TrinityCore/TrinityCore#28971] [Q&A] [answered]
Most of the documentation is only about deployment and very little about development.
The first problem I encountered when trying to develop on the Master branch was that the quest dialog would not close automatically after completing a quest. I'm trying to fix this but I can't.
So what is the exact process for fixing a bug? I found the `Player::RewardQuest` function but I can't find any code to control the quest dialog.
I wonder how the core was developed in the first place? Does it grab the request from the client?

**Comment by mdX7** [comment, marked as answer]
> The first problem I encountered when trying to develop on the Master branch was that the quest dialog would not close automatically after completing a quest

This is likely caused by a missing packet which currently isn't sent to the client after it clicked "complete quest". 
To grab the data i'd suggest using a sniffer (we provide https://github.com/trinitycore/ymir ) on retail servers, then parse the sniff using WowPacketParser (wpp): https://github.com/trinitycore/wowpacketparser. Then you can also enable packet log for your local worldserver in worldserver.conf with the field `PacketLogFile` and do the exact same quest stuff, to be as close as possible to the thing you just did on retail. This also generates a sniff file which is parsable by wpp. You should then take a careful look at all the related `_QUEST_` packets, e.g. `CMSG_QUEST_GIVER_COMPLETE_QUEST`, `SMSG_QUEST_GIVER_QUEST_COMPLETE`. CMSG means ClientMessage (=> sent by client), SMSG being ServerMessage respectively (=> sent by server)
Then you can compare both packet logs (the human readable `<sniffname>_parsed.txt` generated by wpp) and figure out which packet is missing to be sent from out worldserver.

> I wonder how the core was developed in the first place? Does it grab the request from the client?

It's developed based on reverse engineering of the client and sniffing data from retail servers

**Reply by HuangStomach** [reply]
Wow, thank you for your reply, it was very enlightening. It does make more sense and is well documented to develop through sniffing.
